tsl/src/compression/compression.c

```diff
@@ -298,14 +320,232 @@ compress_chunk(Oid in_table, Oid out_table, const ColumnCompressionInfo **column
 
 	TupleDesc in_desc = RelationGetDescr(in_rel);
 	TupleDesc out_desc = RelationGetDescr(out_rel);
+	in_rel_index_oids = RelationGetIndexList(in_rel);
+	int i = 0;
+	/* We create index on the compressed chunk based on ColumnCompressionInfo,
+	 * So, before calling row compressor we need to sort the rows as per ColumnCompressionInfo.
+	 * Cost of sorting can be mitigated if we find and existing BTREE index defined for
+	 * uncompressed chunk otherwise expensive tuplesort will come into play.
+	 *
+	 * This interesting piece of code is trying to find any existing index that
+	 * matches the ColumnCompressionInfo just to override tuplesort with indexscan.
+	 *
+	 * Matching Criteria for Each IndexAtt[i] and ColumnCompressionInfo Keys[i]
+	 * ========================================================================
+	 * a) Index attnum must match with ColumnCompressionInfo Key {keys[i]}.
+	 * b) Index attOption(ASC/DESC and NULL_FIRST) can be mapped with ColumnCompressionInfo
+	 * orderby_asc and null_first.
+	 *
+	 * BTREE Indexes Ordering
+	 * =====================
+	 * a) ASC[Null_Last] ==> [1]->[2]->NULL
+	 * b) [Null_First]ASC ==> NULL->[1]->[2]
+	 * c) DSC[Null_Last]  ==> [2]->[1]->NULL
+	 * d) [Null_First]DSC ==> NULL->[2]->[1]
+	 */
+	foreach (lc, in_rel_index_oids)
+	{
+		Oid index_oid = lfirst_oid(lc);
+		Relation index_rel = index_open(index_oid, AccessShareLock);
+		IndexInfo *index_info = BuildIndexInfo(index_rel);
+		int previous_direction = NoMovementScanDirection;
+		int current_direction = NoMovementScanDirection;
 
-	Tuplesortstate *sorted_rel = compress_chunk_sort_relation(in_rel, n_keys, keys);
-
-	RowCompressor row_compressor;
+		if (n_keys <= index_info->ii_NumIndexKeyAttrs && index_info->ii_Am == BTREE_AM_OID)
+		{
+			for (i = 0; i < n_keys; i++)
+			{
+				int16 att_num = get_attnum(in_table, NameStr(keys[i]->attname));
+				int16 option = index_rel->rd_indoption[i];
+				bool is_orderby_asc = keys[i]->orderby_asc;
+				bool is_null_first = keys[i]->orderby_nullsfirst;
+				if (index_info->ii_IndexAttrNumbers[i] == att_num)
+				{
+					if (COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]))
+					{
+						if (option)
+						{
+							if ((option & INDOPTION_DESC) != 0)
+							{
+								current_direction = BackwardScanDirection;
+							}
+							else
+							{
+								current_direction = ForwardScanDirection;
+							}
+						}
+						else
+						{
+							current_direction = ForwardScanDirection;
+						}
+					}
+					else if (COMPRESSIONCOL_IS_ORDER_BY(keys[i]))
+					{
+						if (option)
+						{
+							/* Case NullFirst(DESC) e.g. [NULL->2->1]*/
+							if (((option & INDOPTION_DESC) != 0) &&
+								((option & INDOPTION_NULLS_FIRST) != 0) &&
+								(is_orderby_asc && is_null_first))
+							{
+								current_direction = BackwardScanDirection;
+							}
```

Comment:
We can simplify these branches as discussed on standup:

```
bool index_orderby_asc = (option & INDOPTION_DESC) == 0;
bool index_nulls_first = (option & INDOPTION_NULLS_FIRST) == 0;

if (is_orderby_asc = index_orderby_asc && is_nulls_fist = index_nulls_first)
{
    current_direction = ForwardScanDirection;
}
else if (is_orderby_asc != index_orderby_asc && is_nulls_first != index_nulls_first)
{
    current_direction = BackwardScanDirection;
}
else
{
    // doesn't match.
}
```

Reply:
Looks good so far, it only misses couple of corner case like this, but I think its good to go with your simplification. let me check after making the changes.
`if (((option & INDOPTION_DESC) != 0) &&
								((option & INDOPTION_NULLS_FIRST) != 0) &&
								(is_orderby_asc && is_null_first))
							{
								current_direction = BackwardScanDirection;
							}`